### ISSUE TITLE
Replace asserts with NC_EINTERNAL...

### ISSUE THREAD

**Issue by DennisHeimbigner** [Unidata/netcdf-c#859]
I think we should consider replacing as many asserts as possible with a return value
of NC_EINTERNAL. Crashing the library is generally not a good idea from the user
point of view.

**Comment by edhartnett** [comment]
I used to feel this way, then Russ convinced me to give asserts a try, and I believe they have been very successful in netCDF development. Asserts are only used when we know it cannot happen. And I don't believe we are getting a lot of reports of asserts - they are rare.

Right now, when a return happens, it is because a return is possible. When an assert happens, we have made a bad programming error. At that point, crashing is preferable to continuing on and maybe messing up data files.

Asserts are well-understood and do improve code quality. I believe they should not be removed, but if someone wants to compile without them, they should do so. (As is routinely done on HPC builds to improve performance).

With the use of test-coverage tools, it is actually pretty easy to be confident that our asserts are truly impossible. Checking for them with if statements is just not needed.

**Comment by DennisHeimbigner** [comment]
My analog would be to ENOMEM. If you are out of memory, you
are largely hosed, but we do not crash, we return an error.

**Comment by edhartnett** [comment]
NOMEM is quite different in that it is possible. 

If I assert(nc) it's because it is impossible for nc to be NULL. If I had to do an if statement instead of an assert, I would just leave it out altogether. 

If we did put in such if-statements, it would be reasonable to protect them with a preprocessor flag, so that they could be left out of compiles.

But that just leaves up re-creating the assert that is already built in. I suggest we just continue to use it.

**Comment by DennisHeimbigner** [comment]
But the fact is that our asserts do get activated. And I suspect that few things
enrage our clients as when the library crashes.

**Comment by edhartnett** [comment]
Do we get a lot of reported asserts?

**Comment by edhartnett** [comment]
I do applaud any effort to improve code quality, but I don't think this is the way to go. @russrew you may enjoy the irony of seeing me make the same arguments you used to me about 15 years ago. ;-)

I agree that if a user sees an assert, that is a very serious bug that should be addressed immediately. But it is rare. Looking at the 101 open issues, none of them are from asserts. When I look at closed issues, I can see only one that seems to be a user hitting an assert.

As you know I have just done an extensive round of code review, extra testing with test coverage analysis, and refactoring of the libsrc4 code. I am confident that asserts are being used correctly there and would not like to see that changed. In fact, I checked all asserts to make sure we are asserting the maximum we can assert, and checking the minimum with if statements. (This is for performance purposes on HPC.)

If you wish to return error instead, then the way to do this would be to provide your own assert function. This way asserts could be used in a standard way in the code, and existing asserts would not need to change. Your assert function could print an error message, and then return an error instead of halting the process. Your assert function, like the C library one, must go away when the library is compiled with 
NDEBUG, so that users can compile without asserts as they do now.

In practice, this will not be an improvement, IMO. Much user code ignores return codes, unfortunately. Even when they check them, the usual result is to then halt the program. Very few science users have code sophisticated enough to deal with such an error in any better way than to just exit. Certainly climate and weather models, written in Fortran, do not have sophisticated error recovery code.

Anyway, what else would be safe to do? If you hit an assert on some libsrc4 code, it means you are missing a necessary pointer. How can anything proceed? Obviously the library is hosed and must be restarted to get it's internal data back into a useful state.

Here's a very typical use of asserts in the libsrc4 code (note that this code is only called for netCDF-4 files):

```
   /* Find info for this file and group, and set pointer to each. */
   if ((retval = nc4_find_nc_grp_h5(ncid, &nc, &grp, &h5)))
      return retval;
   assert(nc && grp && h5);
```

In order for the assert to fail, we have called a libsrc4 function on a non-netcdf-4 file (because there is no h5 or grp) or we have no file info struct (nc). There is no reason to continue. If memory is this hosed up, then calling another netCDF-4 function is not a good idea.

So, in summary, asserts have proven valuable to many C projects, and they are valuable in netCDF. Error conditions that can occur should never be checked by an assert, so users should rarely see them. When users do see them, it is hard to think of how a returned error is more useful than just exiting.

**Comment by DennisHeimbigner** [comment]
Wrong question. The correct question is: do we get a lot of reports of a crash?

**Comment by edhartnett** [comment]
I guess I don't understand your question. Perhaps you could elaborate.

**Comment by DennisHeimbigner** [comment]
I do not recall ever receiving a support question that said
that an assert failed. I have, however, received support questions
that said "netcdf crashed" and it turned out to be an assert failure.
Basically most users are not savy enough to report an assert failure.
They only report "it crashed".

**Comment by edhartnett** [comment]
If they included the error message, it would have included the word "assert" which is what I searched for.

I believe most of the crashes were due to bad parameter checking which has been recently improved. For example the rename functions would crash if you passed NULL for the names. Now they return NC_EINVAL, as they should.

Anyway, I guess I have made my case for asserts as well as I can. This matches my usual advice, which is do what all the other well-behaved C projects do.

**Comment by DennisHeimbigner** [comment]
Didi you search e-support?

**Comment by edhartnett** [comment]
By e-support do you mean the github issues? That is what I searched. I don't have access to any other support systems you guys might be using.

**Comment by DennisHeimbigner** [comment]
Sorry, I forgot that you can not access the esupport (which is the same as
when you worked here).

**Comment by edwardhartnett** [comment]
I hope that I have talked @DennisHeimbigner out of this idea. If so, this issue may be closed.

**Comment by WardF** [comment]
@DennisHeimbigner, what do you think?

**Comment by DennisHeimbigner** [comment]
Its ok to close this, but Ed and I will have to agree to disagree.

**Comment by edwardhartnett** [comment]
OK, seems like this issue can be closed.

**Comment by edwardhartnett** [comment]
OK, seems like this issue can be closed.
